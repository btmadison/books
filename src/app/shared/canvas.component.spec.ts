import { fakeAsync } from '@angular/core/testing';
import { Spectator, createComponentFactory } from '@ngneat/spectator/jest';
import { CanvasComponent } from './canvas.component';
import { ngMocks } from 'ng-mocks';
import { ElementRef } from '@angular/core';

// [9] testing @ViewChild, ngMocks.findInstance(ElementRef) is an effective technique to access @ViewChild

// Note: To make canvas work, had to add an npm package jest-canvas-mock
// Also had to add a specific import to our src/setupTests.ts file: import 'jest-canvas-mock'

describe('Canvas component', () => {
  let component: CanvasComponent;
  let spectator: Spectator<CanvasComponent>;
  let mockSampleCanvas;

  const createComponent = createComponentFactory({
    component: CanvasComponent,
    detectChanges: false
  });

  (window as any).__zone_symbol__FakeAsyncTestMacroTask = [
    {
      source: 'HTMLCanvasElement.toBlob',
      callbackArgs: [{size: 200}],
    },
  ];

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
  });

  it('sanity', fakeAsync(() => {
    spectator.detectChanges();

    expect(component.blobSize).toBe(0);

    spectator.tick();
    expect(component.blobSize).toBeGreaterThan(0);
    expect(spectator.fixture).toMatchSnapshot();
  }));

  it('should check child component @ViewChild', () => {
    spectator.detectChanges();
    mockSampleCanvas = ngMocks.findInstance(ElementRef); // KEY effective technique to access @ViewChild
    // mockSampleCanvas.nativeElement // if you wanted to, you could modify the child component based on your test needs

    expect(mockSampleCanvas.nativeElement).toHaveAttribute('id', 'root1');
    expect(mockSampleCanvas.nativeElement).toBeTruthy().toMatchSnapshot();
  });
});
