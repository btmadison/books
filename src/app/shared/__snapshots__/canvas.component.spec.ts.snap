// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas component sanity 1`] = `
<sample-canvas
  __ngContext__={[Function LRootView]}
  blobSize={[Function Number]}
  sampleCanvas={[Function ElementRef]}
>
  <canvas
    height="200"
    width="200"
  />
</sample-canvas>
`;

exports[`Canvas component should check child component @ViewChild 1`] = `
<div
  id="root1"
>
  <canvas
    height="200"
    width="200"
  />
</div>
`;
